price_service/server-rust/src/network/rpc.rs

```diff
@@ -0,0 +1,118 @@
+use {
+    crate::{
+        network::p2p::OBSERVATIONS,
+        Vaa,
+    },
+    anyhow::Result,
+    axum::{
+        routing::get,
+        Router,
+    },
+    dashmap::DashMap,
+    std::sync::Arc,
+};
+
+mod rest;
+
+#[derive(Clone, Default)]
+pub struct VaaCache(Arc<DashMap<String, Vec<(i64, Vaa)>>>);
+
+impl VaaCache {
+    /// Add a VAA to the cache. Keeps the cache sorted by timestamp.
+    fn add(&mut self, key: String, timestamp: i64, vaa: Vaa) -> Result<()> {
+        self.remove_expired()?;
+        let mut entry = self.0.entry(key).or_default();
+        let key = entry
+            .binary_search_by(|(t, _)| t.cmp(&timestamp))
+            .unwrap_or_else(|e| e);
+        entry.insert(key, (timestamp, vaa));
+        Ok(())
+    }
+
+    /// Remove expired VAA's from the cache.
+    fn remove_expired(&mut self) -> Result<()> {
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)?
+            .as_secs() as i64;
```

Comment:
i think you need to subtract a fixed duration here (right now this is going to clear the cache entirely every time it runs)

Reply:
That's correct, actually all the code in this is just temporary and bad and I just wanted something that fulfilled the interface, I wrote this with Ali quite quickly just so I could understand the current price service. I'll update this whole API to be make more sense once I understand what Ali is doing with the API.